frcxx_nft/src/lib.rs

```diff
@@ -179,7 +250,16 @@ where
         let caller = self.msg.resolve_id(caller)?;
         let operator = self.msg.resolve_or_init(operator)?;
 
-        self.state.approve_for_tokens(&self.bs, caller, operator, token_ids)?;
+        self.transaction(|state, bs| {
+            let mut token_array = state.get_token_data_amt(bs)?;
+            NFTState::assert_owns_tokens(&token_array, caller, token_ids)?;
+
+            state.approve_for_tokens(&mut token_array, operator, token_ids)?;
```

Comment:
I think you can use a predicate over token data here too, to avoid looking it up twice. Similar for some other methods below – I won't comment them all.